src/main.cc

```diff
@@ -923,19 +923,9 @@ mainReconfigureFinish(void *)
     if (Config2.onoff.enable_purge)
         Config2.onoff.enable_purge = 2;
 
-    // parse the config returns a count of errors encountered.
     const int oldWorkers = Config.workers;
-    try {
-        if (parseConfigFile(ConfigFile) != 0) {
-            // for now any errors are a fatal condition...
-            self_destruct();
-        }
-    } catch (...) {
-        // for now any errors are a fatal condition...
-        debugs(1, DBG_CRITICAL, "FATAL: Unhandled exception parsing config file. " <<
-               " Run squid -k parse and check for errors.");
-        self_destruct();
-    }
+
+    (void) parseConfigFile(ConfigFile); // handles any config problems or dies
```

Comment:
maybe it could be clearer text. I am trying to document why the (void) is used.

Right now self_destruct() calls abort(). In future the throw you desire would "skip out" / "unwind the stack" back out to some higher level in main.cc. Either way parseConfigFile() *cannot* return to this reconfigure caller if there are any errors the parser could not already recover from.

Reply:
> Right now self_destruct() calls abort().

To be more accurate, self_destruct() leads to exit(), but yes.

> In future the throw you desire would "skip out" / "unwind the stack" back out to some higher level in main.cc.

Yes, self_destruct() (or equivalent -- self_destruct() should be eventually removed!) will throw in the future.

This PR should polish/fix/add/etc. code that catches that future exception. Since some parsing code is already throwing exceptions, there is no need to wait for that future to come -- from the catching code point of view, the general need and specific use cases are already here! When you are fixing the official catching code (and it does require fixing!), you may assume that self_destruct() already throws -- that assumption (or the lack of thereof) should not affect the catching code fixed by this PR.

Where should the parsing exceptions be caught? I suspect that they should be caught in several places:

* this PR: The high-level parsing code (to provide parsing context and aggregate multiple errors);
* this PR: SquidMainSafe() in main.cc (to terminate the program);
* future PR: mainReconfigureFinish() in main.cc (to skip the failed reconfiguration attempt).

We may adjust this sketch as we go, of course, but this is where I would start.

> Either way parseConfigFile() cannot return to this reconfigure caller if there are any errors the parser could not already recover from.

The "to this reconfigure caller" part is a potentially dangerous red herring IMO: parseConfigFile()'s return behavior should not depend on the caller context. parseConfigFile() should not return (to any caller) on errors.

Starting with this PR, parseConfigFile() should throw on errors. It does not yet do that in this PR, and the PR code this change request is attached to is an illustration of why the current approach does not work well. It is not just the lack of "clarity" in the description or comments. The return type itself is wrong -- it should be void instead of boolean. Besides changing the return type, other related changes in the PR will also be necessary, of course, but this is a good starting point IMO.